packages/suite/src/components/wallet/WalletLayout/components/CoinjoinAccountDiscovery/useAccountLoadingProgress.ts

```diff
@@ -0,0 +1,87 @@
+/* eslint-disable no-nested-ternary */
+
+import { useEffect, useReducer } from 'react';
+import { ScanProgressInfo } from '@trezor/coinjoin/lib/types/backend';
+
+import { CoinjoinService } from 'src/services/coinjoin';
+import { selectSelectedAccount } from 'src/reducers/wallet/selectedAccountReducer';
+import { useSelector } from 'src/hooks/suite/useSelector';
+import { getAccountProgressHandle } from 'src/utils/wallet/coinjoinUtils';
+import { TranslationKey } from 'src/components/suite/Translation';
+
+const INIT_THRESHOLD = 0.05;
+const MEMPOOL_THRESHOLD = 0.85;
+
+type ProgressInfoReducerState = {
+    messageId?: TranslationKey;
+    progress: number;
+    stage?: 'block' | 'mempool';
+    outOf?: { current: number; total: number };
+};
+
+const progressInfoReducer = (
+    state: ProgressInfoReducerState,
+    info: ScanProgressInfo,
+): ProgressInfoReducerState => {
+    const outOf = !info.progress
+        ? info.stage === state.stage
+            ? state.outOf
+            : undefined
+        : info.stage === 'mempool'
+        ? info.progress
+        : {
+              current: info.progress.current - info.progress.from,
+              total: info.progress.to - info.progress.from,
+          };
+
+    const progress = outOf
+        ? info.stage === 'block'
+            ? INIT_THRESHOLD + (outOf.current / outOf.total) * (MEMPOOL_THRESHOLD - INIT_THRESHOLD)
+            : MEMPOOL_THRESHOLD + (outOf.current / outOf.total) * (1 - MEMPOOL_THRESHOLD)
+        : state.stage === 'block' && info.stage === 'mempool'
+        ? MEMPOOL_THRESHOLD
+        : info.stage === 'block'
+        ? INIT_THRESHOLD
+        : state.progress;
+
+    const messageId = `TR_COINJOIN_DISCOVERY_${info.stage === 'block' ? 'BLOCK' : 'MEMPOOL'}_${
+        outOf ? 'PROGRESS' : 'FETCHING'
+    }` as const;
+
+    return {
+        messageId,
+        progress,
+        stage: info.stage,
+        outOf,
+    };
+};
+
+export const useAccountLoadingProgress = () => {
+    const selectedAccount = useSelector(selectSelectedAccount);
+    const [progressInfo, dispatchProgressInfo] = useReducer(progressInfoReducer, {
```

Comment:
Well, `ScanProgressInfo` IS the action in this context :innocent:

The thing is that I need both incoming `ScanProgressInfo` and current `ProgressInfoReducerState` to calculate the next state. So I can

a) transform `state` and `incomingData` together manually and then set it by `setState`, or
b) use the arrow function inside `setState` to transform the previous `state` together with `incomingData`,

which is exactly what `useReducer` does out-of-the-box. Moreover, this way I don't have to have any more helper function in the `useEffect` deps (or defined inside the `useEffect` itself), because whole dispatch function is stable.

Reply:
Yes, let's start using `useReducer` for message-based flows with an object state like this :)